### ISSUE TITLE
Bugs with sql expressions

### ISSUE THREAD

**Issue by brendanheywood** [catalyst/moodle-tool_dataflows#194]
Bugs:
* [x] We should be able to use variables in sql which are unconditional, ie the variable must exist and so we do not need the [[ ]] around them
* [x] I don't think it accepts [[ ]] spread out over multiple lines
* [x] Add a paragraph under the sql field with some examples of sql with vars
* [x] fix issues with some expressions breaking the admin ui and you can't fix it

**Comment by keevan** [comment]
Can you provide examples alongside the reported issue please? I think it'd be good to have the various formats included in tests.

**Comment by brendanheywood** [comment]
The problem is so far I've tried a whole bunch of things and so far zero has worked and they are all different errors. Just found yet another one which can completely bricked the admin ui and I can't edit the step anymore!

Exception - Maximum function nesting level of '256' reached, aborting!

I had tried:

```
SELECT generate_series AS num
FROM generate_series(1, 5);
```
```
SELECT generate_series AS num
FROM generate_series(${{num}}, ${{num}} + 3);
```

```
SELECT generate_series AS num
FROM generate_series([[${{num}}]], [[${{num}}]] + 3);
```
```
SELECT generate_series AS num
[[
FROM generate_series(${{num}}, ${{num}} + 3);
]]
```

**Comment by brendanheywood** [comment]
This bricked the whole thing: (raw config from sql)

```
sql: "SELECT generate_series AS num\r\nFROM generate_series([[${{steps.users.config.num}}]], [[${{steps.users.config.num}}]] + 3)\r\n"
counterfield: num
countervalue: '3'
```

**Comment by keevan** [comment]
Hmm I think the current pattern matching (regex) doesn't support multiline matches. 

Thanks for those examples, I'll see to fixing those at some point probably tomorrow.

**Comment by keevan** [comment]
> This bricked the whole thing: (raw config from sql)
> 
> ```
> sql: "SELECT generate_series AS num\r\nFROM generate_series([[${{steps.users.config.num}}]], [[${{steps.users.config.num}}]] + 3)\r\n"
> counterfield: num
> countervalue: '3'
> ```

Can you provide an export of the dataflow itself? It has various bits that would help to quickly retest the scenario.

**Comment by keevan** [comment]
Oh, I also realised the variable you used doesn't technically exist. You'll need to use `countervalue` literally.

What you probably need to change it to would be:
```
SELECT generate_series AS num
FROM generate_series(${{countervalue}}, ${{countervalue}} + 3);
```
The countervalue will be updated to the value of the previous generate_series value after the iterator has be executed. The query itself, will only be re-queried the next time the flow is entered using the new countervalue.

`counterfield` indicates what key it pulls from the initial record to store in countervalue.

An example dataflow I've adjusted - with the **bug 1** fixed is configured as follows:
```
sql: "SELECT generate_series AS num\r\nFROM generate_series(${{countervalue}}, ${{countervalue}} + 3);"
counterfield: num
countervalue: '6'
```
[sql_20220609_0847.yml.txt](https://github.com/catalyst/moodle-tool_dataflows/files/8868590/sql_20220609_0847.yml.txt)

The relevant commit with the fix https://github.com/catalyst/moodle-tool_dataflows/commit/11323f6ebbe182c5a0a39ba0579000c062c7e586
